File: encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/store/Store.java
package com.automattic.encryptedlogging.store;

import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import com.automattic.encryptedlogging.Dispatcher;
import com.automattic.encryptedlogging.annotations.action.Action;

public abstract class Store {
    protected final Dispatcher mDispatcher;

    Store(Dispatcher dispatcher) {
        mDispatcher = dispatcher;
        mDispatcher.register(this);
    }
    public class OnChanged {}

    /**
     * onAction should {@link Subscribe} with ASYNC {@link ThreadMode}.
     */
    public abstract void onAction(Action action);
    public abstract void onRegister();

    protected void emitChange(OnChanged onChangedEvent) {
        mDispatcher.emitChange(onChangedEvent);
    }
}

File: encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/store/Store.kt
package com.automattic.encryptedlogging.store

import com.automattic.encryptedlogging.Dispatcher

internal abstract class Store internal constructor(private val mDispatcher: Dispatcher) {
    init {
        mDispatcher.register(this)
    }

    /**
     * onAction should [org.greenrobot.eventbus.Subscribe] with ASYNC [org.greenrobot.eventbus.ThreadMode].
     */
    abstract fun onAction()
    abstract fun onRegister()
    protected fun emitChange(onChangedEvent: OnChanged<*>) {
        mDispatcher.emitChange(onChangedEvent)
    }
}
